README.md

```diff
@@ -0,0 +1,203 @@
+├── azure_region 
+├── azure_resource_prefix 
+├── azure_resourcegroup 
+├── azure_subscription_id 
+└── azure_tenant_id
+```
+
+or as a Kubernetes secret defined as:
+
+```
+apiVersion: v1
+data:
+  azure_client_id: NGYwM2JjYWItN2I2My00SSoaNjE3LTk2NDEtMGUyZWViOWNjNWVi
+  azure_client_secret: aEs3TVUzMaiajiDD2tfXy0uVGNYaDU1LVpOcm5FMHMtUndpskD4ueFBvQg==
+  azure_region: ZWFzdHVz
+  azure_resource_prefix: Y2ktbG4tNnNqMzMydCs0xZDA5ZC1kZHE2Mg==
+  azure_resourcegroup: Y2ktbG4tNnNqMzMydC0xZDA5ZC1kZHE2Mi1yZw==
+  azure_subscription_id: ZDM4ZjFlMzgsokPSAAHUDH
+  azure_tenant_id: NjA0DSJAaN2M3ZTktYjJhZsokAOKADH
+kind: Secret
+metadata:
+  name: cloud-credentials
+  namespace: openshift-cloud-network-config-controller
+type: Opaque
+```
+
+# Attributes affecting assignments - subnets / capacity / NICs
+
+Assigning private IP addresses to instances on the cloud comes with some
+limitations.
+
+## Subnets
+
+An IP address can only be assigned to subnets which can host it. In the cloud
+subnets vary depending on the availability zone a instances is located in. This
+controller thus also retrieves the subnet information from the cloud API and
+annotates the node object with it, see below for what this annotation looks
+like.
+
+## Capacity
+
+Clouds limit the amount of private IP addresses which can be associated with
+instances. Some have a variable capacity per instance type and IP family (AWS)
+others have a global and fix capacity defined (GCP/Azure). This controller *does
+not* validate that if capacity is superseded or not, it is up to the client
+controller to initialize its state and track how many assignments are still
+possible.
+
+# NICs
+
+Any CloudPrivateIPConfig currently is only added to the instances' first NIC in
+the order specified by the cloud. On Azure the notion of a "primary" instance
+exists, and is hence used, even if that NIC might not be defined first. As to
+account for future work where IP addresses might be assigned to other NICs
+besides the first one, the annotation reports an array of
+interface/subnet/capacity. For now this array is always of length 1, with the
+first/"primary" interface.
+
+All of these attributes are placed on the node object as an annotation, which
+looks like:
+
+```
+cloud.network.openshift.io/egress-ipconfig: [{"interface": "$IFNAME/$IFID", "ifaddr": {"ipv4": "$IPv4_ADDRESS/$IPv4_SUBNET_MASK", "ipv6": "$IPv6_ADDRESS/$IPv6_SUBNET_MASK"}, "capacity": {"ip": "$IPv4_AND_IPv6_CAPACITY"}}]
+```
+
+if the capacity is IP family agnostic. If that is not the case, the annotation
+will look like:
+
+```
+cloud.network.openshift.io/egress-ipconfig: [{"interface": "$IFNAME/$IFID", "ifaddr": {"ipv4": "$IPv4_ADDRESS/$IPv4_SUBNET_MASK", "ipv6": "$IPv6_ADDRESS/$IPv6_SUBNET_MASK"}, "capacity": {"ipv4": "$IPv4_CAPACITY", "ipv6": "$IPv6_CAPACITY"}}]
+```
+
+# How to hack/debug
+
+When debugging this component make sure to do the following:
+
+1. Run the `./hack/run-locally.sh`, but make sure you have a `KUBECONFIG`
```

Comment:
I looked at the script and main.go and am a little lost on what it does. Can you write a short blurb here with an example of how you would use it? Maybe to set the node annotation; or to use our own CloudPrivateIPConfig yaml?

Reply:
This script doesn't actually help a developer test anything depending on the outcome of the cloud API, ex: verifying that an IP can get assigned/removed, getting the information needed for the annotation it sets on the node...this is because you won't be able to connect to the cloud API endpoints from your DEV environment. What it however allows is you debugging the core control loop functionality. I will mention this.

Reply:
With some minor [fixes](https://github.com/openshift/cloud-network-config-controller/pull/21) I am able to use CNCC in full, at least on GCP